File: Backend/src/main/java/com/etour/controller/TourDetailController.java
package com.etour.controller;

import java.util.List;

import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.web.bind.annotation.*;

import com.etour.dto.TourDetailDtos.ItineraryDto;
import com.etour.dto.TourDetailDtos.JourneyDetailDto;
import com.etour.dto.TourDetailDtos.StayMealDto;
import com.etour.dto.TourDetailDtos.TourAddonDto;
import com.etour.dto.TourDetailDtos.TourContentDto;
import com.etour.dto.TourDetailDtos.TourMediaDto;
import com.etour.service.TourDetailService;

import jakarta.validation.Valid;

/**
 * Everything shown on the "Tour Page" tabs from the BRD (Journey, Stay &
 * Meals, Passport/Visa, Weather, Do's & Don'ts, Terms, Gallery/Video/Map,
 * Add-ons) lives under the owning tour's URL. Reads are public (a visitor
 * browsing a tour page shouldn't need to log in); writes are admin-only.
 */
@RestController
@RequestMapping("/api/tours/{tourId}")
public class TourDetailController {

    private final TourDetailService service;

    public TourDetailController(TourDetailService service) {
        this.service = service;
    }

    // ----- Journey -----
    @GetMapping("/journey")
    public ResponseEntity<List<JourneyDetailDto>> getJourney(@PathVariable Long tourId) {
        return ResponseEntity.ok(service.getJourney(tourId));
    }

    @PostMapping("/journey")
    @PreAuthorize("hasRole('ADMIN')")
    public ResponseEntity<JourneyDetailDto> addJourneyLeg(@PathVariable Long tourId,
            @Valid @RequestBody JourneyDetailDto dto) {
        return ResponseEntity.status(HttpStatus.CREATED).body(service.addJourneyLeg(tourId, dto));
    }

    // ----- Stay & Meals -----
    @GetMapping("/stay-meals")
    public ResponseEntity<List<StayMealDto>> getStayMeals(@PathVariable Long tourId) {
        return ResponseEntity.ok(service.getStayMeals(tourId));
    }

    @PostMapping("/stay-meals")
    @PreAuthorize("hasRole('ADMIN')")
    public ResponseEntity<StayMealDto> addStayMeal(@PathVariable Long tourId, @Valid @RequestBody StayMealDto dto) {
        return ResponseEntity.status(HttpStatus.CREATED).body(service.addStayMeal(tourId, dto));
    }

    @DeleteMapping("/stay-meals/{stayMealId}")
    @PreAuthorize("hasRole('ADMIN')")
    public ResponseEntity<Void> deleteStayMeal(@PathVariable Long tourId, @PathVariable Long stayMealId) {
        service.deleteStayMeal(tourId, stayMealId);
        return ResponseEntity.noContent().build();
    }

    // ----- Content tabs (Passport/Visa, Weather, Do's & Don'ts, Terms) -----
    @GetMapping("/content")
    public ResponseEntity<List<TourContentDto>> getContent(@PathVariable Long tourId) {
        return ResponseEntity.ok(service.getContent(tourId));
    }

    @PostMapping("/content")
    @PreAuthorize("hasRole('ADMIN')")
    public ResponseEntity<TourContentDto> upsertContent(@PathVariable Long tourId,
            @Valid @RequestBody TourContentDto dto) {
        return ResponseEntity.status(HttpStatus.CREATED).body(service.upsertContent(tourId, dto));
    }

    @DeleteMapping("/content/{tourContentId}")
    @PreAuthorize("hasRole('ADMIN')")
    public ResponseEntity<Void> deleteContent(@PathVariable Long tourId, @PathVariable Long tourContentId) {
        service.deleteContent(tourId, tourContentId);
        return ResponseEntity.noContent().build();
    }

    // ----- Media (Gallery / Video / Map / Brochure) -----
    @GetMapping("/media")
    public ResponseEntity<List<TourMediaDto>> getMedia(@PathVariable Long tourId) {
        return ResponseEntity.ok(service.getMedia(tourId));
    }

    @PostMapping("/media")
    @PreAuthorize("hasRole('ADMIN')")
    public ResponseEntity<TourMediaDto> addMedia(@PathVariable Long tourId, @Valid @RequestBody TourMediaDto dto) {
        return ResponseEntity.status(HttpStatus.CREATED).body(service.addMedia(tourId, dto));
    }

    @DeleteMapping("/media/{mediaId}")
    @PreAuthorize("hasRole('ADMIN')")
    public ResponseEntity<Void> deleteMedia(@PathVariable Long tourId, @PathVariable Long mediaId) {
        service.deleteMedia(tourId, mediaId);
        return ResponseEntity.noContent().build();
    }

    // ----- Add-ons -----
    @GetMapping("/addons")
    public ResponseEntity<List<TourAddonDto>> getAddons(@PathVariable Long tourId) {
        return ResponseEntity.ok(service.getAddons(tourId));
    }

    @PostMapping("/addons")
    @PreAuthorize("hasRole('ADMIN')")
    public ResponseEntity<TourAddonDto> addAddon(@PathVariable Long tourId, @Valid @RequestBody TourAddonDto dto) {
        return ResponseEntity.status(HttpStatus.CREATED).body(service.addAddon(tourId, dto));
    }

    @PutMapping("/addons/{addonId}")
    @PreAuthorize("hasRole('ADMIN')")
    public ResponseEntity<TourAddonDto> updateAddon(@PathVariable Long tourId, @PathVariable Long addonId,
            @Valid @RequestBody TourAddonDto dto) {
        return ResponseEntity.ok(service.updateAddon(tourId, addonId, dto));
    }

    @DeleteMapping("/addons/{addonId}")
    @PreAuthorize("hasRole('ADMIN')")
    public ResponseEntity<Void> deleteAddon(@PathVariable Long tourId, @PathVariable Long addonId) {
        service.deleteAddon(tourId, addonId);
        return ResponseEntity.noContent().build();
    }

    // ----- Itinerary -----
    @GetMapping("/itinerary")
    public ResponseEntity<List<ItineraryDto>> getItinerary(@PathVariable Long tourId) {
        return ResponseEntity.ok(service.getItinerary(tourId));
    }

    @PostMapping("/itinerary")
    @PreAuthorize("hasRole('ADMIN')")
    public ResponseEntity<ItineraryDto> addItineraryDay(@PathVariable Long tourId,
            @Valid @RequestBody ItineraryDto dto) {
        return ResponseEntity.status(HttpStatus.CREATED).body(service.addItineraryDay(tourId, dto));
    }

    @PutMapping("/itinerary/{itineraryId}")
    @PreAuthorize("hasRole('ADMIN')")
    public ResponseEntity<ItineraryDto> updateItineraryDay(@PathVariable Long tourId,
            @PathVariable Long itineraryId, @Valid @RequestBody ItineraryDto dto) {
        return ResponseEntity.ok(service.updateItineraryDay(tourId, itineraryId, dto));
    }

    @DeleteMapping("/itinerary/{itineraryId}")
    @PreAuthorize("hasRole('ADMIN')")
    public ResponseEntity<Void> deleteItineraryDay(@PathVariable Long tourId, @PathVariable Long itineraryId) {
        service.deleteItineraryDay(tourId, itineraryId);
        return ResponseEntity.noContent().build();
    }
}

File: dotnet-backend/src/eTour.Api/Controllers/TourDetailController.cs
using eTour.Application.Dtos;
using eTour.Application.Services;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;

namespace eTour.Api.Controllers;

/// <summary>Everything on the "Tour Page" tabs (Journey, Stay &amp; Meals, content tabs, Gallery/Video/Map, Add-ons, Itinerary) lives under the owning tour's URL.</summary>
[ApiController]
[Route("api/tours/{tourId:long}")]
public class TourDetailController : ControllerBase
{
    private readonly ITourDetailService _service;

    public TourDetailController(ITourDetailService service)
    {
        _service = service;
    }

    // ----- Journey -----
    [HttpGet("journey")]
    [AllowAnonymous]
    public async Task<ActionResult<IReadOnlyList<JourneyDetailDto>>> GetJourney(long tourId, CancellationToken ct) =>
        Ok(await _service.GetJourneyAsync(tourId, ct));

    [HttpPost("journey")]
    [Authorize(Roles = "ADMIN")]
    public async Task<ActionResult<JourneyDetailDto>> AddJourneyLeg(long tourId, [FromBody] JourneyDetailDto dto, CancellationToken ct) =>
        StatusCode(StatusCodes.Status201Created, await _service.AddJourneyLegAsync(tourId, dto, ct));

    // ----- Stay & Meals -----
    [HttpGet("stay-meals")]
    [AllowAnonymous]
    public async Task<ActionResult<IReadOnlyList<StayMealDto>>> GetStayMeals(long tourId, CancellationToken ct) =>
        Ok(await _service.GetStayMealsAsync(tourId, ct));

    [HttpPost("stay-meals")]
    [Authorize(Roles = "ADMIN")]
    public async Task<ActionResult<StayMealDto>> AddStayMeal(long tourId, [FromBody] StayMealDto dto, CancellationToken ct) =>
        StatusCode(StatusCodes.Status201Created, await _service.AddStayMealAsync(tourId, dto, ct));

    [HttpDelete("stay-meals/{stayMealId:long}")]
    [Authorize(Roles = "ADMIN")]
    public async Task<IActionResult> DeleteStayMeal(long tourId, long stayMealId, CancellationToken ct)
    {
        await _service.DeleteStayMealAsync(tourId, stayMealId, ct);
        return NoContent();
    }

    // ----- Content tabs -----
    [HttpGet("content")]
    [AllowAnonymous]
    public async Task<ActionResult<IReadOnlyList<TourContentDto>>> GetContent(long tourId, CancellationToken ct) =>
        Ok(await _service.GetContentAsync(tourId, ct));

    [HttpPost("content")]
    [Authorize(Roles = "ADMIN")]
    public async Task<ActionResult<TourContentDto>> UpsertContent(long tourId, [FromBody] TourContentDto dto, CancellationToken ct) =>
        StatusCode(StatusCodes.Status201Created, await _service.UpsertContentAsync(tourId, dto, ct));

    [HttpDelete("content/{tourContentId:long}")]
    [Authorize(Roles = "ADMIN")]
    public async Task<IActionResult> DeleteContent(long tourId, long tourContentId, CancellationToken ct)
    {
        await _service.DeleteContentAsync(tourId, tourContentId, ct);
        return NoContent();
    }

    // ----- Media -----
    [HttpGet("media")]
    [AllowAnonymous]
    public async Task<ActionResult<IReadOnlyList<TourMediaDto>>> GetMedia(long tourId, CancellationToken ct) =>
        Ok(await _service.GetMediaAsync(tourId, ct));

    [HttpPost("media")]
    [Authorize(Roles = "ADMIN")]
    public async Task<ActionResult<TourMediaDto>> AddMedia(long tourId, [FromBody] TourMediaDto dto, CancellationToken ct) =>
        StatusCode(StatusCodes.Status201Created, await _service.AddMediaAsync(tourId, dto, ct));

    [HttpDelete("media/{mediaId:long}")]
    [Authorize(Roles = "ADMIN")]
    public async Task<IActionResult> DeleteMedia(long tourId, long mediaId, CancellationToken ct)
    {
        await _service.DeleteMediaAsync(tourId, mediaId, ct);
        return NoContent();
    }

    // ----- Add-ons -----
    [HttpGet("addons")]
    [AllowAnonymous]
    public async Task<ActionResult<IReadOnlyList<TourAddonDto>>> GetAddons(long tourId, CancellationToken ct) =>
        Ok(await _service.GetAddonsAsync(tourId, ct));

    [HttpPost("addons")]
    [Authorize(Roles = "ADMIN")]
    public async Task<ActionResult<TourAddonDto>> AddAddon(long tourId, [FromBody] TourAddonDto dto, CancellationToken ct) =>
        StatusCode(StatusCodes.Status201Created, await _service.AddAddonAsync(tourId, dto, ct));

    [HttpPut("addons/{addonId:long}")]
    [Authorize(Roles = "ADMIN")]
    public async Task<ActionResult<TourAddonDto>> UpdateAddon(long tourId, long addonId, [FromBody] TourAddonDto dto, CancellationToken ct) =>
        Ok(await _service.UpdateAddonAsync(tourId, addonId, dto, ct));

    [HttpDelete("addons/{addonId:long}")]
    [Authorize(Roles = "ADMIN")]
    public async Task<IActionResult> DeleteAddon(long tourId, long addonId, CancellationToken ct)
    {
        await _service.DeleteAddonAsync(tourId, addonId, ct);
        return NoContent();
    }

    // ----- Itinerary -----
    [HttpGet("itinerary")]
    [AllowAnonymous]
    public async Task<ActionResult<IReadOnlyList<ItineraryDto>>> GetItinerary(long tourId, CancellationToken ct) =>
        Ok(await _service.GetItineraryAsync(tourId, ct));

    [HttpPost("itinerary")]
    [Authorize(Roles = "ADMIN")]
    public async Task<ActionResult<ItineraryDto>> AddItineraryDay(long tourId, [FromBody] ItineraryDto dto, CancellationToken ct) =>
        StatusCode(StatusCodes.Status201Created, await _service.AddItineraryDayAsync(tourId, dto, ct));

    [HttpPut("itinerary/{itineraryId:long}")]
    [Authorize(Roles = "ADMIN")]
    public async Task<ActionResult<ItineraryDto>> UpdateItineraryDay(long tourId, long itineraryId, [FromBody] ItineraryDto dto, CancellationToken ct) =>
        Ok(await _service.UpdateItineraryDayAsync(tourId, itineraryId, dto, ct));

    [HttpDelete("itinerary/{itineraryId:long}")]
    [Authorize(Roles = "ADMIN")]
    public async Task<IActionResult> DeleteItineraryDay(long tourId, long itineraryId, CancellationToken ct)
    {
        await _service.DeleteItineraryDayAsync(tourId, itineraryId, ct);
        return NoContent();
    }
}
